src/main/java/com/zendesk/maxwell/schema/MysqlSavedSchema.java

```diff
@@ -414,69 +418,111 @@ private void restoreSchemaMetadata(Connection conn, Long schemaID) throws SQLExc
+
+			// Table
+			String tName = rs.getString("tableName");
+			String tCharset = rs.getString("tableCharset");
+			String tPKs = rs.getString("tablePk");
+
+			// Column
+			String columnName = rs.getString("columnName");
+			int columnLengthInt = rs.getInt("columnLength");
+			String columnEnumValues = rs.getString("columnEnumValues");
+			String columnCharset = rs.getString("columnCharset");
+			String columnType = rs.getString("columnColtype");
+			int columnIsSigned = rs.getInt("columnIsSigned");
+
+			if (currentDatabase == null || !currentDatabase.getName().equals(dbName)) {
+				if (currentDatabase != null) {
+					long end = System.currentTimeMillis();
+					LOGGER.debug("Processed database " + dbName + " after " + (end - dbStart) + "ms");
+				}
+				dbStart = System.currentTimeMillis();
+				currentDatabase = new Database(dbName, dbCharset);
+				this.schema.addDatabase(currentDatabase);
+				LOGGER.debug("Processing database " + dbName);
+			}
 
-	private void restoreTable(Connection connection, Database d, String name, int id, String charset, String pks) throws SQLException {
-		Statement s = connection.createStatement();
+			if (tName == null) {
+				// if tName is null, there are no tables connected to this database
+				continue;
+			} else if (currentTable == null || !currentTable.getName().equals(tName)) {
+				currentTable = currentDatabase.buildTable(tName, tCharset);
+				if (tPKs != null) {
+					List<String> pkList = Arrays.asList(StringUtils.split(tPKs, ','));
+					currentTable.setPKList(pkList);
+				}
+				columnIndex = 0;
+			}
 
-		Table t = d.buildTable(name, charset);
 
-		ResultSet cRS = s.executeQuery("SELECT * from `columns` where table_id = " + id + " ORDER by id");
+			if (columnName == null) {
+				// If columnName is null, there are no columns connected to this table
+				continue;
+			}
 
-		int i = 0;
-		while (cRS.next()) {
 			Long columnLength;
-			int columnLengthInt = cRS.getInt("column_length");
-			if ( cRS.wasNull() )
+			if (rs.wasNull()) {
 				columnLength = null;
-			else
+			} else {
 				columnLength = Long.valueOf(columnLengthInt);
+			}
 
 			String[] enumValues = null;
-			if ( cRS.getString("enum_values") != null )
-				enumValues = StringUtils.splitByWholeSeparatorPreserveAllTokens(cRS.getString("enum_values"), ",");
+			if (columnEnumValues != null) {
+				enumValues = StringUtils.splitByWholeSeparatorPreserveAllTokens(columnEnumValues, ",");
+			}
 
 			ColumnDef c = ColumnDef.build(
-					cRS.getString("name"), cRS.getString("charset"),
-					cRS.getString("coltype"), i++,
-					cRS.getInt("is_signed") == 1,
+					columnName,
+					columnCharset,
+					columnType,
+					columnIndex++,
+					columnIsSigned == 1,
 					enumValues,
-					columnLength);
-			t.addColumn(c);
-		}
+					columnLength
+			);
+			currentTable.addColumn(c);
 
-		if ( pks != null ) {
-			List<String> pkList = Arrays.asList(StringUtils.split(pks, ','));
-			t.setPKList(pkList);
 		}
+		rs.close();
```

Comment:
oh, gotcha, the objects are added as they're seen and get filled out as the loop progresses.  ok.